### ISSUE TITLE
ICE in ArrayType::calldataEncodedSize(bool) [Array size does not fit unsigned]

### ISSUE THREAD

**Issue by agroce** [ethereum/solidity#8382]
## Description

This contract:
```solidity
contract C {
    struct R { uint[10][10] y; }
    struct S { uint a; uint b; R d; uint[20][20][2999999999999999999999999990] c; }
    function f() public {
        C.S memory y;
        C.S[10] memory z;        y.a < 2;
    }
}
```

produces:

```
Internal compiler error during compilation:
/home/user/solidity/libsolidity/ast/Types.cpp(1662): Throw in function virtual unsigned int solidity::frontend::ArrayType::calldataEncodedSize(bool) const
Dynamic exception type: boost::exception_detail::clone_impl<solidity::langutil::InternalCompilerError>
std::exception::what: Array size does not fit unsigned.
[solidity::util::tag_comment*] = Array size does not fit unsigned.
```

when compiled with solc

On master, using AFL fuzzing.  Another discovery using https://github.com/agroce/afl-compiler-fuzzer, now with a few more mutation operators

## Environment

- Compiler version:  0.6.3-develop.2020.2.10+commit.e8eb1f2d.Linux.clang
- Target EVM version (as per compiler settings): N/A
- Framework/IDE (e.g. Truffle or Remix): N/A
- EVM execution environment / backend / blockchain client: N/A
- Operating system: Ubuntu 18.04 in docker

## Steps to Reproduce

Above shows pretty clearly, I think.

**Comment by bshastry** [comment]
This might be a duplicate of #8286 that may be solved by #8291

**Comment by bshastry** [comment]
Thank you @agroce for bringing this to our attention.

**Comment by bshastry** [comment]
Not exactly a duplicate of #8286 since the assertion that fails is this

https://github.com/ethereum/solidity/blob/44bcff42f5fa835dbda83c9f6120f858038bb0f7/libsolidity/ast/Types.cpp#L1664

**Comment by bshastry** [comment]
Reopening because although #8665 fixed the original issue, the following new test case throws the same ICE

```
pragma experimental ABIEncoderV2;
contract C {
  struct S { T[] u; }
  struct T { int[333333333] x; }
  function f() public returns (S memory) {}
}
```

throws

https://github.com/ethereum/solidity/blob/5f8299b379a1f4d99f491a0b91fbbda853e580e5/libsolidity/ast/Types.cpp#L1742

**Comment by bshastry** [comment]
A smaller test input for this issue

```
pragma experimental ABIEncoderV2;

contract C {
  int[277777777] i;
  function f() external { abi.encode(i); }
}
```

**Comment by cameel** [comment]
Just for reference, the two examples above cause the same assertion but through different paths. The first one is when return value gets implicitly ABI-encoded and the second one is an explicit call to `abi.encode()`.

Here's a simplified code that goes through the same path as the first example:
```solidity
pragma experimental ABIEncoderV2;

contract C {
    struct S { int[1000000000] v; }
    function f() public returns (S[1] memory) {}
}
```

**Comment by cameel** [comment]
Looks like fixing this will be more invasive than I originally thought. The underlying issue is that most of the functions on the `Type` class that compute sizes (i.e. `calldataEncodedSize()`, `calldataHeadSize()`, `calldataEncodedTailSize()` and `memoryHeadSize()`) return `unsigned`. And that's enough for valid code but if we want to display a nice message for stuff that's too large we run into this assertion inside them.

In memory and in calldata it happens only for structs because the code that calculates the size works around the `unsigned` limitation by finding the underlying array element type (going through all the nesting levels) and multiplying by length in all dimensions inside `bigint`. It assumes that the element size always fits in `unsigned`, which was probably true before structs were introduced but now it's not.

This is why this example results in a compiler error (`Error: Type too large for memory.`):
```solidity
contract C {
    function f() internal {
        uint[1000000000] memory x;
    }
}
```
but this one compiles just fine:
```solidity
contract D {
    struct S { uint[1000000000000000000000000000000000000000000000000000000000000000000] v; }
    function f() internal {
        S[1] memory x;
    }
}
```

The most straightforward fix would be to change the return values of these functions from `unsigned` to `bigint` or `u256` but the extra overhead seems pointless if that's only for invalid contracts.

An alternative solution would be to make the function that calculates size take structs into account but I'm afraid that we'll eventually just end up with two parallel systems for computing sizes and the one without the size limit will keep breaking because we won't remember to update it when adding new types just like it happened there.

**Comment by cameel** [comment]
I'm unassigning myself since I'm not actively working on it and it's blocked by and old refactor PR (#9746). I intend to go back to this at some point though unless someone else takes it in the meantime.

**Comment by cameel** [comment]
Another snippet triggering the same issue (from #11609):
```solidity
contract C {
    struct S {
        bytes32[1263941234127518272] a;
    }

    function f(S[] memory) public {}
}
```

This is almost the same thing as the one from my comment above (https://github.com/ethereum/solidity/issues/8382#issuecomment-686695040).

**Comment by cameel** [comment]
This will probably be fixed by the work related to compile-time constant evaluation so we won't be working on this specifically in the near future.
